e_os.h

```diff
@@ -296,20 +296,17 @@ static ossl_inline void ossl_sleep(unsigned long millis)
     ts.tv_sec = (long int) (millis / 1000);
     ts.tv_nsec = (long int) (millis % 1000) * 1000000ul;
     nanosleep(&ts, NULL);
-# elif defined(__TANDEM)
-#  if !defined(_REENTRANT)
+# elif defined(__TANDEM) && !defined(_REENTRANT)
 #   include <cextdecs.h(PROCESS_DELAY_)>
+
     /* HPNS does not support usleep for non threaded apps */
     PROCESS_DELAY_(millis * 1000);
-#  elif defined(_SPT_MODEL_)
-#   include <spthread.h>
-#   include <spt_extensions.h>
-    usleep(millis * 1000);
-#  else
-    usleep(millis * 1000);
-#  endif
 # else
-    usleep(millis * 1000);
+    unsigned int s = (unsigned int)(millis / 1000);
+    unsigned int us = (unsigned int)((millis % 1000) * 1000);
+
+    sleep(s);
```

Comment:
Interactions between sleep() and any of ... or usleep() are unspecified.

Reply:
Also IMO `sleep()` should not be called with s == 0, otherwise this will cause an unneeded additional yield of cpu.

Reply:
FWIW, this is exactly how OSSL_sleep is implemented in master currently.  If its a problem here, then it seems like its a problem more generally

Reply:
Hmm... right, it should be fixed there as well. Not a serious issue of course, but anyway.

Reply:
If the goal is to protect sleep and usleep from zero values:
```
    unsigned int s = (unsigned int)(millis / 1000);
    unsigned int us = (unsigned int)((millis % 1000) * 1000);

    if (s > 0)
        sleep(s);
    if (us > 0)
        usleep(us);
```
However, I would caution that this could change the threading behaviour. Both `sleep()` and `usleep()` will potentially cause a context switch between threads. The current implementation ensures that there are two calls (as has been the case for 3.3 in OSSL_sleep(). Changing that might change the scheduling of execution of the threads, particularly if OSSL_sleep(0) is called.

Reply:
@rsbeckerca for that reason I propose to add only the `s > 0` check for `sleep()` call. I'd call usleep() unconditionally. If we are asking to sleep for at least 1 second I do not think it matters whether there are two yields of the CPU. However for smaller sleep values it might make a difference.

Reply:
Related to the above, it might be useful (if not already existing) to have a threadtest that calls `OSSL_sleep(0)` and ensures an detects a thread swap. That would detect sleep(0) behaviour. `OSSL_sleep(1000)` already deals with calling `usleep(0)` in master.

Reply:
Can we handle that as a separate Issue and I can do the two PRs needed for separate approval?

Reply:
> Related to the above, it might be useful (if not already existing) to have a threadtest that calls `OSSL_sleep(0)` and ensures an detects a thread swap. That would detect sleep(0) behaviour. `OSSL_sleep(1000)` already deals with calling `usleep(0)` in master.

I think it would be hard to test this reliably.

Reply:
> > Related to the above, it might be useful (if not already existing) to have a threadtest that calls `OSSL_sleep(0)` and ensures an detects a thread swap. That would detect sleep(0) behaviour. `OSSL_sleep(1000)` already deals with calling `usleep(0)` in master.
> 
> I think it would be hard to test this reliably.

I rather agree. Since the existing `OSSL_sleep()` code now tests stable, we should keep it and apply the changes in older branches.

Reply:
@nhorman Do you want an update to this to add the `if (s > 0)` code? What about the current master?

Reply:
@rsbeckerca please update this PR and create another PR against master to add `if (s > 0)` there as well.

Reply:
> @rsbeckerca please update this PR and create another PR against master to add `if (s > 0)` there as well.

Will do. PR #24012 covers master and 3.3.

Reply:
@t8m PR #23972 updated as requested.